src/Symfony/Component/Translation/Bridge/Loco/Provider/LocoProvider.php

```diff
@@ -0,0 +1,245 @@
+use Symfony\Component\Translation\TranslatorBagInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Mathieu Santostefano <msantostefano@protonmail.com>
+ *
+ * In Loco:
+ *  * Tags refers to Symfony's translation domains
+ *  * Assets refers to Symfony's translation keys
+ *  * Translations refers to Symfony's translated messages
+ *
+ * @experimental in 5.3
+ */
+final class LocoProvider implements ProviderInterface
+{
+    private $client;
+    private $loader;
+    private $logger;
+    private $defaultLocale;
+    private $endpoint;
+
+    public function __construct(HttpClientInterface $client, LoaderInterface $loader, LoggerInterface $logger, string $defaultLocale, string $endpoint)
+    {
+        $this->client = $client;
+        $this->loader = $loader;
+        $this->logger = $logger;
+        $this->defaultLocale = $defaultLocale;
+        $this->endpoint = $endpoint;
+    }
+
+    public function __toString(): string
+    {
+        return sprintf('%s://%s', LocoProviderFactory::SCHEME, $this->endpoint);
+    }
+
+    public function getName(): string
+    {
+        return LocoProviderFactory::SCHEME;
+    }
+
+    public function write(TranslatorBagInterface $translatorBag): void
+    {
+        $catalogue = $translatorBag->getCatalogue($this->defaultLocale);
+
+        if (!$catalogue) {
+            $catalogue = $translatorBag->getCatalogues()[0];
+        }
+
+        // Create keys on Loco
+        foreach ($catalogue->all() as $domain => $messages) {
+            $ids = [];
+            foreach ($messages as $id => $message) {
+                $ids[] = $id;
+                $this->createAsset($id);
+            }
+            if ($ids) {
+                $this->tagsAssets($ids, $domain);
+            }
+        }
+
+        // Push translations in all locales and tag them with domain
+        foreach ($translatorBag->getCatalogues() as $catalogue) {
+            $locale = $catalogue->getLocale();
+
+            if (!\in_array($locale, $this->getLocales())) {
+                $this->createLocale($locale);
+            }
+
+            foreach ($catalogue->all() as $messages) {
+                foreach ($messages as $id => $message) {
+                    $this->translateAsset($id, $message, $locale);
+                }
+            }
+        }
+    }
+
+    public function read(array $domains, array $locales): TranslatorBagInterface
+    {
+        $domains = $domains ?? ['*'];
+        $translatorBag = new TranslatorBag();
+
+        foreach ($locales as $locale) {
+            foreach ($domains as $domain) {
+                $response = $this->client->request('GET', sprintf('export/locale/%s.xlf?filter=%s', $locale, $domain));
+                $responseContent = $response->getContent(false);
+
+                if (200 !== $response->getStatusCode()) {
+                    throw new ProviderException('Unable to read the Loco response: '.$responseContent, $response);
```

Comment:
If we are trying to read a locale that does not exists. We get the following exception: 

```
In LocoProvider.php line 106:
                                                                            
  Unable to read the Loco response: <?xml version="1.0" encoding="UTF-8"?>  
  <loco>                                                                    
    <status>404</status>                                                    
    <error>Locale not found</error>                                         
  </loco>                                                                   
                                                                            
```

I suggest we add `$locale` and `$domain` in the exception message. Maybe event customise if for 404?